### PR TITLE
Add support to Checks API

### PR DESCRIPTION
**Pull request by sileht**
We was supporting only Statuses API for required_status_check/contexts.

But Github have recently added a new Checks API. And Travis now send
Check Runs and Check Suites to Github in additional to Statuses.

As Github support now Check Runs names for required_status_check/contexts,
we should to support it, too.

To do this we have a couple of things to do:
* Looks at Check Runs to compute out "required_statuses_succeed"
* Subscribes and handles some check_suite/check_run Github events to
  refresh the Mergify status

This change does all of that.

### CODE REVIEW

**Comment by sileht** [comment]
We can't merge it yet, see https://github.com/Mergifyio/mergify-engine/pull/102/files#diff-c20883d43327ab2e60651352ef187b4eR103

I have reached the Github support about these two issues. The new API is not really reliable for us.

I will propose a doc change to explicitly say we didn't support the Checks API because of Github bugs.

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 70948753f1848c7c5c3e25c33aebf192fe063dbb

File: mergify_engine/check_api.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        headers={'Accept': 'application/vnd.github.antiope-preview+json'}
+    )
+    return data
+
+
+@tenacity.retry(wait=tenacity.wait_exponential(multiplier=0.2),
+                stop=tenacity.stop_after_attempt(5),
+                retry=tenacity.retry_never)
+def workaround_for_unfinished_check_suite(g_repo, data):
+    """Workaround for broken Checks API events
```
Comment (jd) [inline mergify_engine/check_api.py]: final dot missing

File: mergify_engine/check_api.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    The Checks API have two major flaws:
```
Comment (jd) [inline mergify_engine/check_api.py]: has

File: mergify_engine/check_api.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    * We received check_suite/completed too early, some checks are still
+      pending
+    * Once it's completed, even if some check are rerun, we never got
```
Comment (jd) [inline mergify_engine/check_api.py]: checks

File: mergify_engine/check_api.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      a new event to notify us that it finish again.
```
Comment (jd) [inline mergify_engine/check_api.py]: finished

File: mergify_engine/check_api.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+
+    This method workaround the first issue, but the second one is still
```
Comment (jd) [inline mergify_engine/check_api.py]: works around, while the second one

File: mergify_engine/check_api.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    unsolved
```
Comment (jd) [inline mergify_engine/check_api.py]: final dot missing

**Comment by sileht** [comment]
Pull-request updated, HEAD is now f2eee8a44bf63b13052d7aeed97ecc2f25d80cb9

**Comment by sileht** [comment]
@jd I think we should merge this. The Github "Checks API" is not yet perfect for us, but we should at least compute our "CI combined statuses" correctly. Some CI applications have started to use it, and unlike Travis they haven't done a transition phase with both API.

For example: https://github.com/probot/dco

File: mergify_engine/check_api.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                stop=tenacity.stop_after_attempt(5),
+                retry=tenacity.retry_never)
+def workaround_for_unfinished_check_suite(g_repo, data):
+    """Workaround for broken Checks API events.
+
+    The Checks API have two major flaws:
+    * We received check_suite/completed too early, some checks are still
+    pending
+    * Once it's completed, even if some check are rerun, we never got
+    a new event to notify us that it finish again.
```
Comment (jd) [inline mergify_engine/check_api.py]: finishes

**Pull request merged**